zone/attack.cpp

```diff
@@ -5388,6 +5388,106 @@ void Mob::TryPetCriticalHit(Mob *defender, DamageHitInfo &hit)
 	}
 }
 
+bool Mob::RollMeleeCritCheck(Mob *defender, EQ::skills::SkillType skill)
+{
+	// We either require an innate crit chance or some SPA 169 to crit
+	bool innate_crit = false;
+	int crit_chance = GetCriticalChanceBonus(skill);
+	// Paladin check
+	if (defender->IsUndeadForSlay()) {
+		crit_chance = crit_chance + GetUndeadSlayRate();
+	}
+
+	if ((GetClass() == Class::Warrior || GetClass() == Class::Berserker) && GetLevel() >= 12) {
+		innate_crit = true;
+	} else if (GetClass() == Class::Ranger && GetLevel() >= 12 && skill == EQ::skills::SkillArchery) {
+		innate_crit = true;
+	} else if (GetClass() == Class::Rogue && GetLevel() >= 12 && skill == EQ::skills::SkillThrowing) {
+		innate_crit = true;
+	}
```

Comment:
Can consolidate this a bit since they all check `GetLevel() >= 12`.
```cpp
if (GetLevel() >= 12) {
	if (
		GetClass() == Class::Warrior ||
		(GetClass() == Class::Ranger && skill == EQ::skills::SkillArchery) ||
		(GetClass() == Class::Rogue && skill == EQ::skills::SkillThrowing) ||
		GetClass() == Class::Berserker
	) {
		innate_crit = true;
	}
}
```